### PR TITLE
Refactor from_pattern

### PR DESCRIPTION
**Pull request by Tastyep**
Hello,

I'd like to propose a change to the method named `from_pattern`, because it has a big impact I only recoded the function and will wait for your approval before going on.

The motivation is to support all the [Diagnostic](https://github.com/mfussenegger/nvim-lint#from_pattern) attributes and to make the parser flexible regarding what groups will be parsed using the regex.
This also allows adding new diagnostic fields in the future without worrying about refactoring.
Finally, I like the more concise and explicit definition of the parsers, for instance, I tested with cppcheck which could be defined as follows:
```
local pattern = [[([^:]*):(%d*):(%d*): %[([^%]\]*)%] ([^:]*): (.*)]]
local groups = {"file", "lineno", "colno", "code", "severity", "message" }
local severity_map = {
  ["error"] = vim.lsp.protocol.DiagnosticSeverity.Error,
  ["warning"] = vim.lsp.protocol.DiagnosticSeverity.Warning,
  ["performance"] = vim.lsp.protocol.DiagnosticSeverity.Warning,
  ["style"] = vim.lsp.protocol.DiagnosticSeverity.Information,
  ["information"] = vim.lsp.protocol.DiagnosticSeverity.Information,
}
local parser = from_pattern(pattern, groups, severity_map)

```

### CODE REVIEW

**Review by mfussenegger — COMMENTED** [review]
I like the idea :+1:

**Comment by Tastyep** [comment]
I think I'm done with the parser refactoring.
I couldn't test `tidy` and `ruby`, it would be great if you can, I can confirm that the other ones I've modified are working.
I'll now update the [from_pattern](https://github.com/mfussenegger/nvim-lint#from_pattern) section of the README.

**Comment by Tastyep** [comment]
Also, I've removed some logic preventing mypy to not follow imports. While I agree that it makes it verbose, since mypy allows this to be configurable, I don't think that nvim-lint should take responsibility for this.

**Comment by Tastyep** [comment]
I think this is ready.

Regarding my point for mypy, the `file` group is now used to ensure that diagnostics from other files will be filtered out (which was the behavior implemented by some parsers (cppcheck, clang-tidy, mypy)).

File: README.md [hunk 1]
```diff
@@ -127,15 +127,34 @@ The function takes a single argument which is the `errorformat`.
...
-four groups in order: line number, offset, code and message.
+The function allows to parse the linter's output using a lua regex pattern.
 
-The second argument is a skeleton used to create each diagnostic. This can be
-used to provide default values - for example to set a `severity`.
+- pattern: The regex pattern applied on each line of the output
+- groups: The groups specified by the pattern
 
+``` lua
+groups = {"lineno", "message", ("colno" | ("colbeg", "colend")), ["code"], ["codeDesc"], ["file"], ["severity"]}
```
Comment (mfussenegger) [inline README.md:139]:
```suggestion
groups = {"lineno", "message",  "start_col", ["end_col"], ["code"], ["codeDesc"], ["file"], ["severity"]}
```

`start_col` and `end_col` is what's typically used within the neovim code base. I also wouldn't distinguish between `colno` and `start_col`, but just have the former, and default `end_col` to `start_col` if it isn't present.

File: lua/lint/linters/chktex.lua [hunk 2]
```diff
@@ -1,37 +1,35 @@
 local pattern = '(%d+):(%d+):(%d+):(.+):(%d+):(.*)'
 local severities = {
-    Error = vim.lsp.protocol.DiagnosticSeverity.Error,
-    Warning = vim.lsp.protocol.DiagnosticSeverity.Warning,
-    Message = vim.lsp.protocol.DiagnosticSeverity.Information
+  Error = vim.lsp.protocol.DiagnosticSeverity.Error,
+  Warning = vim.lsp.protocol.DiagnosticSeverity.Warning,
```
Comment (mfussenegger) [inline lua/lint/linters/chktex.lua:4]: Please revert the unrelated formatting changes

File: lua/lint/parser.lua [hunk 3]
```diff
@@ -22,53 +21,108 @@ function M.from_errorformat(efm, skeleton)
...
+      end
+
+      local colno = entries['colno'] ~= nil and tonumber(entries['colno']) or 1
+      return {
+        ['start'] = { line = lineno - 1, character = colno - 1 },
+        ['end'] = { line = lineno - 1, character = colno },
+      }
+    end,
+
+    relatedInformation = function(_)
```
Comment (mfussenegger) [inline lua/lint/parser.lua:71]: Why do we already need methods for stuff that is not supported?

File: lua/lint/linters/chktex.lua (hunk 2 above)
Comment (Tastyep) [reply lua/lint/linters/chktex.lua:4]: Sorry about that, I made a small change to the file and reformatted everything.
Do you currently use a formatter so that I can adapt to the same style ?

File: lua/lint/parser.lua (hunk 3 above)
Comment (Tastyep) [reply lua/lint/parser.lua:71]: It was just to show how it should be implemented. Actually, I could just return entries[...] and it would be implemented, but as it was not used I didn't bother :).
Should I has support since it's easy to do ?

File: lua/lint/linters/chktex.lua (hunk 2 above)
Comment (mfussenegger) [reply lua/lint/linters/chktex.lua:4]: I tend to use the formatting that you would've applied, but so far it is not enforced which is why different parsers have a different style. I wouldn't mind the reformat itself - what I dislike is that it reformats *and* makes a change within the same commit.

File: lua/lint/parser.lua (hunk 3 above)
Comment (mfussenegger) [reply lua/lint/parser.lua:71]: I'd remove it if it is currently not used by any parser. We can extend it once needed.

**Comment by Tastyep** [comment]
Thank you for the review, it's indeed better :+1:

**Comment by mfussenegger** [comment]
Thanks again for the PR :+1:

File: lua/lint/parser.lua [hunk 4]
```diff
@@ -22,53 +21,85 @@ function M.from_errorformat(efm, skeleton)
...
+      ['information'] = vim.lsp.protocol.DiagnosticSeverity.Information,
+      ['hint'] = vim.lsp.protocol.DiagnosticSeverity.Hint,
+    }
+  defaults = defaults or {}
+  return function(output, bufnr)
     local diagnostics = {}
+    local buffer_path = vim.api.nvim_buf_get_name(bufnr)
+
+    for _, line in ipairs(vim.fn.split(output, '\n')) do
+      local results = { line:gmatch(pattern)() }
```
Comment (gpanders) [inline lua/lint/parser.lua:78]: Out of curiosity, is there a reason to use `line:gmatch(pattern)()` rather than just `line:match(pattern)`? They do seem to be different, as with the former (current) implementation I am not getting any results from my [nix](https://github.com/mfussenegger/nvim-lint/pull/44) linter, but with the latter implementation it does work.
